src/components/name-tag.js

```diff
@@ -0,0 +1,377 @@
+        },
+        { once: true }
+      );
+      if (this.displayName.length > DISPLAY_NAME_LENGTH) {
+        this.displayName = this.displayName.slice(0, DISPLAY_NAME_LENGTH).concat("...");
+      }
+      this.nametagTextEl.setAttribute("text", {
+        value: this.displayName
+      });
+    }
+
+    if (this.identityName) {
+      if (this.identityName.length > DISPLAY_NAME_LENGTH) {
+        this.identityName = this.identityName.slice(0, DISPLAY_NAME_LENGTH).concat("...");
+      }
+      this.identityNameEl.setAttribute("text", { value: this.identityName });
+      this.identityNameEl.object3D.visible = this.el.sceneEl.is("frozen");
+    }
+  },
+
+  onModelLoading() {
+    this.tmpNametagVisible = false;
+    this.updateNameTagVisibility();
+    this.model = null;
+  },
+
+  onModelLoaded({ detail: { model } }) {
+    clearInterval(this.firstIkStepHandler);
+    const checkIkStep = () => {
+      if (this.ikRootEl.object3D.visible) {
+        clearInterval(this.firstIkStepHandler);
+        this.neckEl = this.el.querySelector(".Neck");
+        this.tmpNametagVisible = true;
+        this.updateNameTagVisibility();
+        this.model = model;
+        this.updateNameTagPosition();
+      }
+    };
+    this.firstIkStepHandler = setInterval(checkIkStep, 500);
+  },
+
+  updateNameTagPosition() {
+    this.updateAvatarModelBBAA();
+    const tmpVector = new THREE.Vector3();
+    this.nametagHeight =
+      Math.abs(tmpVector.subVectors(this.ikRootEl.object3D.position, this.avatarBBAACenter).y) +
+      this.avatarBBAASize.y / 2 +
+      NAMETAG_OFFSET;
+    this.nametagElPosY = this.nametagHeight;
+    this.size = this.nametagTextEl.components["text"].getSize();
+    this.size.x = Math.max(this.size.x, NAMETAG_MIN_WIDTH);
+    this.updateNameTag();
+  },
+
+  onAnalyserVolumeUpdated({ detail: { volume } }) {
+    let isTalking = false;
+    this.volume = volume;
+    this.volumeAvg.push(Date.now(), volume);
+    if (!this.playerInfo.data.muted) {
+      const average = this.volumeAvg.movingAverage();
+      isTalking = average > 0.01;
+    }
+    this.wasTalking = this.isTalking;
+    this.isTalking = isTalking;
+    if (this.nametagVisibility === "showSpeaking") {
+      if (!this.isTalking && this.wasTalking) {
+        this.frozenTimer = setTimeout(() => {
+          this.wasNametagVisible = this.isNametagVisible;
+          this.isNametagVisible = false;
+          this.isNametagVisible !== this.wasNametagVisible && this.updateNameTag();
+        }, 1000);
+      } else if (this.isTalking && !this.wasTalking) {
+        clearTimeout(this.frozenTimer);
+        this.wasNametagVisible = this.isNametagVisible;
+        this.isNametagVisible = true;
+        this.isNametagVisible !== this.wasNametagVisible && this.updateNameTag();
+      }
+    }
+    this.isNametagVisible && this.isTalking !== this.wasTalking && this.updateBorder();
+    this.isNametagVisible && this.updateVolume();
+  },
+
+  onStateChanged() {
```

Comment:
This wrapping function feels superfluous.

Reply:
Updated